File: app/src/main/java/org/mifos/mobile/presenters/ClientChargePresenter.java
package org.mifos.mobile.presenters;

import android.content.Context;

import org.mifos.mobile.R;
import org.mifos.mobile.api.DataManager;
import org.mifos.mobile.injection.ActivityContext;
import org.mifos.mobile.models.Charge;
import org.mifos.mobile.models.Page;
import org.mifos.mobile.presenters.base.BasePresenter;
import org.mifos.mobile.ui.views.ClientChargeView;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;


/**
 * @author Vishwajeet
 * @since 17/8/16.
 */
public class ClientChargePresenter extends BasePresenter<ClientChargeView> {

    private final DataManager dataManager;
    private CompositeDisposable compositeDisposable;

    /**
     * Initialises the ClientChargePresenter by automatically injecting an instance of
     * {@link DataManager} and {@link Context}.
     *
     * @param dataManager DataManager class that provides access to the data
     *                    via the API.
     * @param context     Context of the view attached to the presenter. In this case
     *                    it is that of an {@link androidx.appcompat.app.AppCompatActivity}
     */

    @Inject
    public ClientChargePresenter(DataManager dataManager, @ActivityContext Context context) {
        super(context);
        this.dataManager = dataManager;
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void attachView(ClientChargeView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        compositeDisposable.clear();
    }

    public void loadClientCharges(long clientId) {
        checkViewAttached();
        getMvpView().showProgress();
        compositeDisposable.add(dataManager.getClientCharges(clientId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<Page<Charge>>() {
                    @Override
                    public void onComplete() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().hideProgress();
                        getMvpView().showErrorFetchingClientCharges(
                                context.getString(R.string.client_charges));
                    }

                    @Override
                    public void onNext(Page<Charge> chargePage) {
                        getMvpView().hideProgress();
                        if (chargePage != null) {
                            getMvpView().showClientCharges(chargePage.getPageItems());
                        }
                    }
                })
        );
    }

    public void loadLoanAccountCharges(long loanId) {
        checkViewAttached();
        getMvpView().showProgress();
        compositeDisposable.add(dataManager.getLoanCharges(loanId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<List<Charge>>() {
                    @Override
                    public void onComplete() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().hideProgress();
                        getMvpView().showErrorFetchingClientCharges(
                                context.getString(R.string.client_charges));
                    }

                    @Override
                    public void onNext(List<Charge> chargeList) {
                        getMvpView().hideProgress();
                        getMvpView().showClientCharges(chargeList);
                    }
                })
        );
    }

    public void loadSavingsAccountCharges(long savingsId) {
        checkViewAttached();
        getMvpView().showProgress();
        compositeDisposable.add(dataManager.getSavingsCharges(savingsId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<List<Charge>>() {
                    @Override
                    public void onComplete() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().hideProgress();
                        getMvpView().showErrorFetchingClientCharges(
                                context.getString(R.string.client_charges));
                    }

                    @Override
                    public void onNext(List<Charge> chargeList) {
                        getMvpView().hideProgress();
                        getMvpView().showClientCharges(chargeList);
                    }
                })
        );
    }

    public void loadClientLocalCharges() {
        checkViewAttached();
        compositeDisposable.add(dataManager.getClientLocalCharges()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<Page<Charge>>() {
                    @Override
                    public void onComplete() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showErrorFetchingClientCharges(
                                context.getString(R.string.client_charges));
                    }

                    @Override
                    public void onNext(Page<Charge> chargePage) {
                        getMvpView().showClientCharges(chargePage.getPageItems());
                    }
                })
        );
    }
}

File: app/src/main/java/org/mifos/mobile/presenters/ClientChargePresenter.kt
package org.mifos.mobile.presenters

import android.content.Context

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers

import org.mifos.mobile.R
import org.mifos.mobile.api.DataManager
import org.mifos.mobile.injection.ActivityContext
import org.mifos.mobile.models.Charge
import org.mifos.mobile.models.Page
import org.mifos.mobile.presenters.base.BasePresenter
import org.mifos.mobile.ui.views.ClientChargeView

import javax.inject.Inject

/**
 * @author Vishwajeet
 * @since 17/8/16.
 */
class ClientChargePresenter @Inject constructor(private val dataManager: DataManager?, @ActivityContext context: Context?) :
        BasePresenter<ClientChargeView?>(context) {

    private val compositeDisposable: CompositeDisposable = CompositeDisposable()
    override fun attachView(mvpView: ClientChargeView?) {
        super.attachView(mvpView)
    }

    override fun detachView() {
        super.detachView()
        compositeDisposable.clear()
    }

    fun loadClientCharges(clientId: Long) {
        checkViewAttached()
        mvpView?.showProgress()
        dataManager?.getClientCharges(clientId)
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribeOn(Schedulers.io())
                ?.subscribeWith(object : DisposableObserver<Page<Charge?>?>() {
                    override fun onComplete() {}
                    override fun onError(e: Throwable) {
                        mvpView?.hideProgress()
                        mvpView?.showErrorFetchingClientCharges(
                                context?.getString(R.string.client_charges))
                    }

                    override fun onNext(chargePage: Page<Charge?>) {
                        mvpView?.hideProgress()
                        mvpView?.showClientCharges(chargePage.pageItems)
                    }
                })?.let {
                    compositeDisposable.add(it
                    )
                }
    }

    fun loadLoanAccountCharges(loanId: Long) {
        checkViewAttached()
        mvpView?.showProgress()
        dataManager?.getLoanCharges(loanId)
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribeOn(Schedulers.io())
                ?.subscribeWith(object : DisposableObserver<List<Charge?>?>() {
                    override fun onComplete() {}
                    override fun onError(e: Throwable) {
                        mvpView?.hideProgress()
                        mvpView?.showErrorFetchingClientCharges(
                                context?.getString(R.string.client_charges))
                    }

                    override fun onNext(chargeList: List<Charge?>) {
                        mvpView?.hideProgress()
                        mvpView?.showClientCharges(chargeList)
                    }
                })?.let {
                    compositeDisposable.add(it
                    )
                }
    }

    fun loadSavingsAccountCharges(savingsId: Long) {
        checkViewAttached()
        mvpView?.showProgress()
        dataManager?.getSavingsCharges(savingsId)
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribeOn(Schedulers.io())
                ?.subscribeWith(object : DisposableObserver<List<Charge?>?>() {
                    override fun onComplete() {}
                    override fun onError(e: Throwable) {
                        mvpView?.hideProgress()
                        mvpView?.showErrorFetchingClientCharges(
                                context?.getString(R.string.client_charges))
                    }

                    override fun onNext(chargeList: List<Charge?>) {
                        mvpView?.hideProgress()
                        mvpView?.showClientCharges(chargeList)
                    }
                })?.let {
                    compositeDisposable.add(it
                    )
                }
    }

    fun loadClientLocalCharges() {
        checkViewAttached()
        dataManager?.clientLocalCharges
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribeOn(Schedulers.io())
                ?.subscribeWith(object : DisposableObserver<Page<Charge?>?>() {
                    override fun onComplete() {}
                    override fun onError(e: Throwable) {
                        mvpView?.showErrorFetchingClientCharges(
                                context?.getString(R.string.client_charges))
                    }

                    override fun onNext(chargePage: Page<Charge?>) {
                        mvpView?.showClientCharges(chargePage.pageItems)
                    }
                })?.let {
                    compositeDisposable.add(it
                    )
                }
    }

}